clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractStickyAssignor.java

```diff
@@ -469,73 +426,184 @@ private boolean allSubscriptionsEqual(Set<String> allTopics,
             String consumer = subscriptionEntry.getKey();
             MemberData memberData = memberData(subscriptionEntry.getValue());
 
-            for (TopicPartition partition: memberData.partitions) {
-                if (sortedPartitionConsumersByGeneration.containsKey(partition)) {
-                    Map<Integer, String> consumers = sortedPartitionConsumersByGeneration.get(partition);
-                    if (memberData.generation.isPresent() && consumers.containsKey(memberData.generation.get())) {
-                        // same partition is assigned to two consumers during the same rebalance.
-                        // log a warning and skip this record
-                        log.warn("Partition '{}' is assigned to multiple consumers following sticky assignment generation {}.",
-                            partition, memberData.generation);
-                    } else
-                        consumers.put(memberData.generation.orElse(DEFAULT_GENERATION), consumer);
-                } else {
-                    TreeMap<Integer, String> sortedConsumers = new TreeMap<>();
-                    sortedConsumers.put(memberData.generation.orElse(DEFAULT_GENERATION), consumer);
-                    sortedPartitionConsumersByGeneration.put(partition, sortedConsumers);
-                }
-            }
-        }
-
-        // prevAssignment holds the prior ConsumerGenerationPair (before current) of each partition
-        // current and previous consumers are the last two consumers of each partition in the above sorted map
-        for (Map.Entry<TopicPartition, TreeMap<Integer, String>> partitionConsumersEntry: sortedPartitionConsumersByGeneration.entrySet()) {
-            TopicPartition partition = partitionConsumersEntry.getKey();
-            TreeMap<Integer, String> consumers = partitionConsumersEntry.getValue();
-            Iterator<Integer> it = consumers.descendingKeySet().iterator();
-
-            // let's process the current (most recent) consumer first
-            String consumer = consumers.get(it.next());
-            currentAssignment.computeIfAbsent(consumer, k -> new ArrayList<>());
-            currentAssignment.get(consumer).add(partition);
-
-            // now update previous assignment if any
-            if (it.hasNext()) {
-                int generation = it.next();
-                prevAssignment.put(partition, new ConsumerGenerationPair(consumers.get(generation), generation));
+            // we already have the maxGeneration info, so just compare the current generation of memberData, and put into prevAssignment
+            if (memberData.generation.isPresent() && memberData.generation.get() < maxGeneration) {
+                // If the current member's generation is lower than maxGeneration, put into prevAssignment if needed
+                updatePrevAssignment(prevAssignment, memberData.partitions, consumer, memberData.generation.get());
+            } else if (!memberData.generation.isPresent()) {
```

Comment:
Since now, we already have ` currentAssignment` from `allSubscriptionsEqual` method, as well as the `maxGeneration` data, so we can simplify the logic here for `prevAssignment` only.